scheds/rust/scx_lavd/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,1749 @@
+		return false;
+	}
+	if (transit_task_stat(taskc, LAVD_TASK_STAT_ENQ))
+		update_stat_for_enq(p, taskc, cpuc);
+
+	/*
+	 * Enqueue the task to the global runqueue based on its virtual
+	 * deadline.
+	 */
+	vdeadline = taskc->eligible_delta_ns + taskc->vdeadline_delta_ns +
+		    bpf_ktime_get_ns();
+	scx_bpf_dispatch_vtime(p, LAVD_GLOBAL_DSQ, LAVD_SLICE_MAX_NS,
+			       vdeadline, enq_flags);
+	return true;
+}
+
+s32 BPF_STRUCT_OPS(lavd_select_cpu, struct task_struct *p, s32 prev_cpu,
+		   u64 wake_flags)
+{
+	bool found_idle = false;
+	s32 cpu_id;
+
+	/*
+	 * When a task wakes up, we should decide where to place the task at
+	 * ops.select_cpu(). We make a best effort to find an idle CPU. If
+	 * there is an idle CPU and a task is a true-wake-up task (not just
+	 * fork-ed or execv-ed), we consider such a task as a latency-critical
+	 * task, so directly dispatch to the local FIFO queue of the chosen
+	 * CPU. If the task is directly dispatched here, the sched_ext won't
+	 * call ops.enqueue().
+	 */
+	cpu_id = scx_bpf_select_cpu_dfl(p, prev_cpu, wake_flags, &found_idle);
+
+	if (found_idle && is_wakeup_wf(wake_flags)) {
+		struct task_ctx *taskc;
+
+		taskc = bpf_task_storage_get(&task_ctx_stor, p, 0, 0);
+		if (!taskc) {
+			scx_bpf_error("task_ctx_stor lookup failed");
+			return prev_cpu;
+		}
+
+		if (!put_local_rq(p, taskc, cpu_id, 0)) {
+			/*
+			 * If a task is overscheduled (greedy_ratio > 1000), we
+			 * do not select a CPU, so that later the enqueue
+			 * operation can put it to the global queue.
+			 */
+			return prev_cpu;
+		}
+	}
+
+	return cpu_id;
+}
+
+void BPF_STRUCT_OPS(lavd_enqueue, struct task_struct *p, u64 enq_flags)
+{
+	struct task_ctx *taskc;
+	
+	/*
+	 * If there is an idle CPU at the ops.select_cpu(), the task is already
+	 * dispatched at ops.select_cpu(), so ops.enqueue() won't be called.
+	 * Hence, the task that is enqueued here are the cases: 1) there is no
+	 * idle CPU when ops.select_cpu() or 2) the task is not the case of
+	 * being wakened up (i.e., resume after preemption). Therefore, we
+	 * always put the task to the global DSQ, so any idle CPU can pick it
+	 * up.
+	 */
+	taskc = bpf_task_storage_get(&task_ctx_stor, p, 0, 0);
+	if (!taskc) {
+		scx_bpf_error("task_ctx_stor lookup failed");
+		return;
+	}
+
+	/*
+	 * Place a task to the global run queue.
+	 */
+	put_global_rq(p, taskc, enq_flags);
+}
+
+void BPF_STRUCT_OPS(lavd_dispatch, s32 cpu, struct task_struct *prev)
+{
+	/*
+	 * Now, a CPU has no task to run, so it gets a task from the global run
+	 * queue for execution.
+	 */
+	scx_bpf_consume(LAVD_GLOBAL_DSQ);
+}
+
+void BPF_STRUCT_OPS(lavd_runnable, struct task_struct *p, u64 enq_flags)
+{
+	struct task_struct *waker;
+	struct task_ctx *taskc;
+	u64 now, interval;
+
+	/*
+	 * When a task @p is wakened up, the wake frequency of its waker task
+	 * is updated. The @current task is a waker and @p is a waiter, which
+	 * is being wakened up now.
+	 */
+	if (!(enq_flags & SCX_ENQ_WAKEUP))
```

Comment:
Maybe `is_wakeup_ef()` is more consistent? I wonder how helpful the `is_wakeup*()` helpers are tho given that they each just test a single flag and used in one or two places.

Reply:
I will fix it as you suggested.

is_wakeup_*() helpers are not very useful. They are mostly for code readability. I wish the code could be read without much knowledge of the kernel flags. As it becomes too cumbersome, I will do refactoring once again.